# Computational Biology

## A quick summary of the contents:

* Lab1: Preperations
<br>

* [Lab2](https://github.com/bijiuni/compu_bio/tree/master/Lab2): Analzying sequences, generating	phylogenetic trees, and identifying sequences using BLAST

* Tools: Biophython, BLAST, matplotlib
<br>

* [Lab3](https://github.com/bijiuni/compu_bio/tree/master/Lab3): Determine statistics on word occurences and Zipf's law

* Tools: matplotlib
<br>

* [Lab4](https://github.com/bijiuni/compu_bio/tree/master/Lab4): Investigate and document the relationships between genes, enzymes and pathways

* Tools: SQL(sqlite3), pandas, Biopython, entrez
<br>

* [Lab5](https://github.com/bijiuni/compu_bio/tree/master/Lab5): Obtain secondary structures of folded RNA investigate riboswitch scenarios in the paper by Penchovsky and Breaker

* Tools: RNAfold
<br>

* [Lab6](https://github.com/bijiuni/compu_bio/tree/master/Lab6): Run bowtie analysis on strains of Shewanella oneidensis, solving contamination and analyzing coverage statistics

* Tools: bowtie2, samtools, matplotlib
<br>

* [Lab7](https://github.com/bijiuni/compu_bio/tree/master/Lab7): Investigate compression algorithms on generated binary and fasta files and real biological information

* Tools: numpy, matplotlib, gzip, bzip2, pbzip2, ArithmeticCompress, entrez
<br>

* [Lab8 & 9](https://github.com/bijiuni/compu_bio/tree/master/Lab8): Assemble genome de novo using PacBio and Illumina reads, extract quality statistics, identify taxonomic origin, and obtain two genome annotations via different pipelines

* Tools: spades, assembly-stats, rna_hmm3, bedtools, SeqMatch, BASys, RAST
