{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Non existent tables. Tables will be created now.\n",
      "Pathway: Glycolysis; EC: 4.1.1.49; Organism: homo sapiens\n",
      "Pathway: Glycolysis; EC: 4.1.1.49; Organism: escherichia coli\n",
      "Pathway: Glycolysis; EC: 4.1.1.49; Organism: drosophila\n",
      "Pathway: Glycolysis; EC: 5.3.1.9; Organism: homo sapiens\n",
      "Pathway: Glycolysis; EC: 5.3.1.9; Organism: escherichia coli\n",
      "Pathway: Glycolysis; EC: 5.3.1.9; Organism: drosophila\n",
      "Pathway: Glycolysis; EC: 5.4.2.2; Organism: homo sapiens\n",
      "Pathway: Glycolysis; EC: 5.4.2.2; Organism: escherichia coli\n",
      "Pathway: Glycolysis; EC: 5.4.2.2; Organism: drosophila\n",
      "Pathway: Glycolysis; EC: 1.8.1.4; Organism: homo sapiens\n",
      "Pathway: Glycolysis; EC: 1.8.1.4; Organism: escherichia coli\n",
      "Pathway: Glycolysis; EC: 1.8.1.4; Organism: drosophila\n",
      "Pathway: TCA cycle; EC: 4.1.1.31; Organism: homo sapiens\n",
      "Pathway: TCA cycle; EC: 4.1.1.31; Organism: escherichia coli\n",
      "Pathway: TCA cycle; EC: 4.1.1.31; Organism: drosophila\n",
      "Pathway: TCA cycle; EC: 6.2.1.4; Organism: homo sapiens\n",
      "Pathway: TCA cycle; EC: 6.2.1.4; Organism: escherichia coli\n",
      "Pathway: TCA cycle; EC: 6.2.1.4; Organism: drosophila\n",
      "Pathway: TCA cycle; EC: 1.1.5.4; Organism: homo sapiens\n",
      "Pathway: TCA cycle; EC: 1.1.5.4; Organism: escherichia coli\n",
      "Pathway: TCA cycle; EC: 1.1.5.4; Organism: drosophila\n",
      "Pathway: TCA cycle; EC: 4.1.1.49; Organism: homo sapiens\n",
      "Pathway: TCA cycle; EC: 4.1.1.49; Organism: escherichia coli\n",
      "Pathway: TCA cycle; EC: 4.1.1.49; Organism: drosophila\n",
      "Pathway: Pentose Phosphate; EC: 3.1.3.11; Organism: homo sapiens\n",
      "Pathway: Pentose Phosphate; EC: 3.1.3.11; Organism: escherichia coli\n",
      "Pathway: Pentose Phosphate; EC: 3.1.3.11; Organism: drosophila\n",
      "Pathway: Pentose Phosphate; EC: 2.7.1.15; Organism: homo sapiens\n",
      "Pathway: Pentose Phosphate; EC: 2.7.1.15; Organism: escherichia coli\n",
      "Pathway: Pentose Phosphate; EC: 2.7.1.15; Organism: drosophila\n",
      "Pathway: Pentose Phosphate; EC: 3.1.1.31; Organism: homo sapiens\n",
      "Pathway: Pentose Phosphate; EC: 3.1.1.31; Organism: escherichia coli\n",
      "Pathway: Pentose Phosphate; EC: 3.1.1.31; Organism: drosophila\n",
      "Pathway: Pentose Phosphate; EC: 5.4.2.7; Organism: homo sapiens\n",
      "Pathway: Pentose Phosphate; EC: 5.4.2.7; Organism: escherichia coli\n",
      "Pathway: Pentose Phosphate; EC: 5.4.2.7; Organism: drosophila\n",
      "\n",
      "Pathways table:\n",
      "   id               name                                        description\n",
      "0   1         Glycolysis  Glycolysis is the process of converting glucos...\n",
      "1   2          TCA cycle  The citrate cycle (TCA cycle, Krebs cycle) is ...\n",
      "2   3  Pentose Phosphate  The pentose phosphate pathway is a process of ...\n",
      "\n",
      "Genes table:\n",
      "    id        name          organism  \\\n",
      "0    1  CD680162.1      homo sapiens   \n",
      "1    2    R86926.1  escherichia coli   \n",
      "2    3  CF389836.1        drosophila   \n",
      "3    4    Z21512.1      homo sapiens   \n",
      "4    5      T44203  escherichia coli   \n",
      "5    6  AW258048.1        drosophila   \n",
      "6    7    X68132.1      homo sapiens   \n",
      "7    8    L39517.1  escherichia coli   \n",
      "8    9    X64011.1        drosophila   \n",
      "9   10    V00898.1      homo sapiens   \n",
      "10  11    U09680.1  escherichia coli   \n",
      "11  12  AG025748.1        drosophila   \n",
      "12  13  BM215258.2      homo sapiens   \n",
      "13  14      T44113  escherichia coli   \n",
      "14  15  CF389472.1        drosophila   \n",
      "15  16    X52763.1      homo sapiens   \n",
      "16  17  AW505992.1  escherichia coli   \n",
      "17  18    X70485.1        drosophila   \n",
      "18  19  BG181451.1      homo sapiens   \n",
      "19  20    R87889.1  escherichia coli   \n",
      "20  21  CF388434.1        drosophila   \n",
      "21  22  CD680162.1      homo sapiens   \n",
      "22  23    R86926.1  escherichia coli   \n",
      "23  24  CF389836.1        drosophila   \n",
      "24  25    Z12967.1      homo sapiens   \n",
      "25  26      T44458  escherichia coli   \n",
      "26  27  AW244174.1        drosophila   \n",
      "27  28    X03616.1      homo sapiens   \n",
      "28  29      T43889  escherichia coli   \n",
      "29  30  AJ390731.1        drosophila   \n",
      "30  31    Z16890.1      homo sapiens   \n",
      "31  32    R87585.1  escherichia coli   \n",
      "32  33  AW258457.1        drosophila   \n",
      "33  34    X60334.1      homo sapiens   \n",
      "34  35      T44637  escherichia coli   \n",
      "35  36  CF388555.1        drosophila   \n",
      "\n",
      "                                             sequence  \n",
      "0   TTCCGCCGGGTGATGGACGGGATCCTGCCGGCGACGAAGTGCAGGA...  \n",
      "1   ATNAAAATAAAGGATATTTTTTTTATTCAAGTAACTGCAAATAGGA...  \n",
      "2   TTGTGACTGAGCTTAAATCGAATGATATAATCGATAATGGCGCCTA...  \n",
      "3   ATCGATTTGAGAGGTGATAGGTGATTAGCGGGGGGAGATGAAAAGT...  \n",
      "4   GAAAAACAGACAGCGGAGGACTAAAACAACTAGCCACAACACAACG...  \n",
      "5   TGGCGGAACCAAGTACAACATTTGAAAAGGACCTGTCAACTGTAGA...  \n",
      "6   TTCCCTGCTCGTATGTCTTGAGATTGGGCCTTAAATAATATGCAGC...  \n",
      "7   ACTTACTACTGTGCCCTCTCTCTTTACTGGGGGATACAGGCACCGA...  \n",
      "8   CGTTATTTAAGGTGTTACATAGTTCTATGGAAATAGGGTCTATACC...  \n",
      "9         TCGAGGAGAAATCGGGCATGGAGGGCCGCAAGAAGATGTTCGA  \n",
      "10  TTATTAACAACGCACTTATCGATCTCCCCGCCCCATCAAACATCTC...  \n",
      "11  AAGCTTGTCTACTTAGAGTCACAGTTTTGAAATTGTAANTGGTATC...  \n",
      "12  GGTTTTATTCCTACAGAAAGTTGATTTTTATGACTGTTTCTAATTT...  \n",
      "13  CCTGANTCGTCTTCTTCGATAATCTCCCTCTTGTTTCAATCAAGAA...  \n",
      "14  AGTGGGCATGCAGGCTTTGTGATGAATTGGTTAAGTGGGCATGCAG...  \n",
      "15  CTCGAGGAGTCACGTGTAATGATGAACTCCAATTAAGAACACAAGT...  \n",
      "16  GAATTCCTTCAATTTCCAAACACAAATAAATCAGCAAAAAACAATG...  \n",
      "17  GTGCAGCTGGTGGAGTCTGGGGGAGGCTTGGTCAAGCCTGGGAGGT...  \n",
      "18  AGAAATTCTGACAGTGTTTGATCCTCCATGCTCGTGTTCTACGCAT...  \n",
      "19  AAACGACCAGAATCCAGGTTCCCAGTTCAGCCAGGCCCTGGCATTC...  \n",
      "20  GTACATTTTTGCAATGATAGGATTGCAAGTAGACTCCAGCTCTTTC...  \n",
      "21  TTCCGCCGGGTGATGGACGGGATCCTGCCGGCGACGAAGTGCAGGA...  \n",
      "22  ATNAAAATAAAGGATATTTTTTTTATTCAAGTAACTGCAAATAGGA...  \n",
      "23  TTGTGACTGAGCTTAAATCGAATGATATAATCGATAATGGCGCCTA...  \n",
      "24  GGTGGAGGAACAAAAGTTTTCGTCCAAGATAAAAACCTTCCTACAG...  \n",
      "25  CAAGTAACATTCTCAGCTTCTAGAAGGTTTTCTCACCAACCCCCAA...  \n",
      "26  GCACTCCGCGGGGACGTTCGACGTGTCGTCGAGGACCGGCGGTCCC...  \n",
      "27  CTCTTATTCTTGTAGATCCTAAACTATTTTGGAAAAGTCAAAATAA...  \n",
      "28  TCAACAAGCTGCATCTCTTACAGAGTTTAGTTGGCTTGGCTACAGA...  \n",
      "29  CCCTCTGGTTTCTCTCGCTTTAGTGGGGCTATTGGTCAGCATCACA...  \n",
      "30  TAAATTAGTGCTTAGTGGGAAATTGTAACTTCAAAAAGTTACATTA...  \n",
      "31  CAATCTGTGTNCCGCGGCCCGGCCTTGNAGGTCTTCCCAGGGTCAG...  \n",
      "32  ACGCGTCCGATCAGATGGCTATTTGTACATGTACAACCTGGACCCC...  \n",
      "33  GAGGTCCAGCTGCAACAGTCTGGACCTGAACTGGTAGAGCCTGGGA...  \n",
      "34  CACAAACAAACAATAACTAAAATGGCTTTGTCTTCCAAAACCCTAA...  \n",
      "35  TATGTTTCAAGCTCTAATTCCAACTAGGAGATTCAAGTTTGATGTT...  \n",
      "\n",
      "Enzymes table:\n",
      "          ec      name                                        description\n",
      "0    5.3.1.9  KYO64656   Glucose-6-phosphate isomerase [Escherichia coli]\n",
      "1    5.4.2.2  PVH54613              phosphoglucomutase [Escherichia coli]\n",
      "2    1.8.1.4  KYO73969     Dihydrolipoyl dehydrogenase [Escherichia coli]\n",
      "3   4.1.1.31  AAA24414  phosphoenolpyruvate carboxylase mutant (EC 4.1...\n",
      "4    6.2.1.4  BAH12424             unnamed protein product [Homo sapiens]\n",
      "5    1.1.5.4  KYO69993   Malate:quinone oxidoreductase [Escherichia coli]\n",
      "6   4.1.1.49  KYO65602  Phosphoenolpyruvate carboxykinase [ATP] [Esche...\n",
      "7   3.1.3.11  KYO73765  Fructose-1,6-bisphosphatase class 1 [Escherich...\n",
      "8   2.7.1.15  CAA20884               EG:115C2.1 [Drosophila melanogaster]\n",
      "9   3.1.1.31  KYO71699       6-phosphogluconolactonase [Escherichia coli]\n",
      "10   5.4.2.7  KYO73479              Phosphopentomutase [Escherichia coli]\n",
      "\n",
      "Pathway enzyme associated table:\n",
      "    id            pathway    enzyme  pathway_order\n",
      "0    1         Glycolysis  4.1.1.49              1\n",
      "1    2         Glycolysis   5.3.1.9              2\n",
      "2    3         Glycolysis   5.4.2.2              3\n",
      "3    4         Glycolysis   1.8.1.4              4\n",
      "4    5          TCA cycle  4.1.1.31              1\n",
      "5    6          TCA cycle   6.2.1.4              2\n",
      "6    7          TCA cycle   1.1.5.4              3\n",
      "7    8          TCA cycle  4.1.1.49              4\n",
      "8    9  Pentose Phosphate  3.1.3.11              1\n",
      "9   10  Pentose Phosphate  2.7.1.15              2\n",
      "10  11  Pentose Phosphate  3.1.1.31              3\n",
      "11  12  Pentose Phosphate   5.4.2.7              4\n",
      "\n",
      "Enzyme genes associated table:\n",
      "    id        ec          organism        gene\n",
      "0    1  4.1.1.49      homo sapiens  CD680162.1\n",
      "1    2  4.1.1.49  escherichia coli    R86926.1\n",
      "2    3  4.1.1.49        drosophila  CF389836.1\n",
      "3    4   5.3.1.9      homo sapiens    Z21512.1\n",
      "4    5   5.3.1.9  escherichia coli      T44203\n",
      "5    6   5.3.1.9        drosophila  AW258048.1\n",
      "6    7   5.4.2.2      homo sapiens    X68132.1\n",
      "7    8   5.4.2.2  escherichia coli    L39517.1\n",
      "8    9   5.4.2.2        drosophila    X64011.1\n",
      "9   10   1.8.1.4      homo sapiens    V00898.1\n",
      "10  11   1.8.1.4  escherichia coli    U09680.1\n",
      "11  12   1.8.1.4        drosophila  AG025748.1\n",
      "12  13  4.1.1.31      homo sapiens  BM215258.2\n",
      "13  14  4.1.1.31  escherichia coli      T44113\n",
      "14  15  4.1.1.31        drosophila  CF389472.1\n",
      "15  16   6.2.1.4      homo sapiens    X52763.1\n",
      "16  17   6.2.1.4  escherichia coli  AW505992.1\n",
      "17  18   6.2.1.4        drosophila    X70485.1\n",
      "18  19   1.1.5.4      homo sapiens  BG181451.1\n",
      "19  20   1.1.5.4  escherichia coli    R87889.1\n",
      "20  21   1.1.5.4        drosophila  CF388434.1\n",
      "21  22  4.1.1.49      homo sapiens  CD680162.1\n",
      "22  23  4.1.1.49  escherichia coli    R86926.1\n",
      "23  24  4.1.1.49        drosophila  CF389836.1\n",
      "24  25  3.1.3.11      homo sapiens    Z12967.1\n",
      "25  26  3.1.3.11  escherichia coli      T44458\n",
      "26  27  3.1.3.11        drosophila  AW244174.1\n",
      "27  28  2.7.1.15      homo sapiens    X03616.1\n",
      "28  29  2.7.1.15  escherichia coli      T43889\n",
      "29  30  2.7.1.15        drosophila  AJ390731.1\n",
      "30  31  3.1.1.31      homo sapiens    Z16890.1\n",
      "31  32  3.1.1.31  escherichia coli    R87585.1\n",
      "32  33  3.1.1.31        drosophila  AW258457.1\n",
      "33  34   5.4.2.7      homo sapiens    X60334.1\n",
      "34  35   5.4.2.7  escherichia coli      T44637\n",
      "35  36   5.4.2.7        drosophila  CF388555.1\n"
     ]
    }
   ],
   "source": [
    "# IMPORTANT: if this document does not compile with your default data rate limit, open jupyter notebook using the \n",
    "# following command: jupyter notebook --NotebookApp.iopub_data_rate_limit=10000000000\n",
    "\n",
    "import sqlite3\n",
    "import pandas as pd\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "import sys\n",
    "\n",
    "\n",
    "Entrez.email = \"thomas.galeon@berkeley.edu\"\n",
    "\n",
    "# Creating 3 SQL tables\n",
    "conn = sqlite3.connect('Lab4.db')\n",
    "c = conn.cursor()\n",
    "\n",
    "# The previous SQL tables are erased if they exist. They will be recreated and populated automatically.\n",
    "try:\n",
    "    c.execute(\"DROP TABLE genes\")\n",
    "    c.execute(\"DROP TABLE pathways\")\n",
    "    c.execute(\"DROP TABLE enzymes\")\n",
    "    c.execute(\"DROP TABLE pathwayenzyme\")\n",
    "    c.execute(\"DROP TABLE enzymegene\")\n",
    "except:\n",
    "    print(\"Non existent tables. Tables will be created now.\")\n",
    "    \n",
    "# 5 SQL Tables are created - the detailed description of these tables can be found in the README file    \n",
    "c.execute(\"\"\"CREATE TABLE genes (id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "                                name TEXT,\n",
    "                                organism TEXT,\n",
    "                                sequence TEXT);\"\"\")\n",
    "\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE pathways (id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "                                    name TEXT,\n",
    "                                    description TEXT);\"\"\")\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE enzymes (ec TEXT PRIMARY KEY,\n",
    "                                    name TEXT,\n",
    "                                    description TEXT);\"\"\")\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE pathwayenzyme (id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "                                            pathway TEXT,\n",
    "                                            enzyme TEXT,\n",
    "                                            pathway_order INTEGER);\"\"\")\n",
    "\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE enzymegene (id INTEGER PRIMARY KEY AUTOINCREMENT,\n",
    "                                            ec TEXT,\n",
    "                                            organism TEXT,\n",
    "                                            gene TEXT);\"\"\")\n",
    "\n",
    "\n",
    "\n",
    "# Considered organisms, pathways and enzymes. Enzymes are given in order of occurence in the pathway.\n",
    "org = ['homo sapiens', 'escherichia coli', 'drosophila']\n",
    "\n",
    "\n",
    "\n",
    "pwt = {'Glycolysis': ['4.1.1.49','5.3.1.9', '5.4.2.2', '1.8.1.4'], \n",
    "      'TCA cycle':['4.1.1.31','6.2.1.4','1.1.5.4','4.1.1.49',],\n",
    "      'Pentose Phosphate': ['3.1.3.11','2.7.1.15', '3.1.1.31','5.4.2.7']}\n",
    "\n",
    "\n",
    "#'TCA cycle':['1.2.4.1', '4.2.1.3', '6.2.1.5', '2.3.1.61'], ['1.3.5.4', '4.1.1.31', '1.1.1.37', '2.3.3.1', '2.3.1.12', '6.4.1.1', '1.1.1.42'\n",
    "#'1.3.5.4', '1.2.4.2', '1.1.1.286', '6.2.1.4', '1.2.7.11' '1.2.4.2',, '6.2.1.4', '1.2.7.11'\n",
    "# The pathway description table is populated manually. The sources for the pathway descriptions are \n",
    "#indicated in the README file.\n",
    "\n",
    "c.execute(\"\"\"INSERT INTO pathways (name,\n",
    "                                    description)\n",
    "                            VALUES\n",
    "                                    ('Glycolysis',\n",
    "                                    'Glycolysis is the process of converting glucose into pyruvate and generating small \n",
    "                                    amounts of ATP (energy) and NADH (reducing power). It is a central pathway that \n",
    "                                    produces important precursor metabolites: six-carbon compounds of glucose-6P and \n",
    "                                    fructose-6P and three-carbon compounds of glycerone-P, glyceraldehyde-3P, \n",
    "                                    glycerate-3P, phosphoenolpyruvate, and pyruvate. Acetyl-CoA, another important \n",
    "                                    precursor metabolite, is produced by oxidative decarboxylation of pyruvate. When the\n",
    "                                    enzyme genes of this pathway are examined in completely sequenced genomes, the \n",
    "                                    reaction steps of three-carbon compounds from glycerone-P to pyruvate form a \n",
    "                                    conserved core module, which is found in almost all organisms and which sometimes\n",
    "                                    contains operon structures in bacterial genomes. Gluconeogenesis is a synthesis \n",
    "                                    pathway of glucose from noncarbohydrate precursors. It is essentially a reversal of \n",
    "                                    glycolysis with minor variations of alternative paths.')\n",
    "                                    ;\"\"\")\n",
    "\n",
    "\n",
    "c.execute(\"\"\"INSERT INTO pathways (name,\n",
    "                                    description)\n",
    "                            VALUES\n",
    "                                    ('TCA cycle',\n",
    "                                    'The citrate cycle (TCA cycle, Krebs cycle) is an important aerobic pathway for the \n",
    "                                    final steps of the oxidation of carbohydrates and fatty acids. The cycle starts with \n",
    "                                    acetyl-CoA, the activated form of acetate, derived from glycolysis and pyruvate \n",
    "                                    oxidation for carbohydrates and from beta oxidation of fatty acids. The two-carbon \n",
    "                                    acetyl group in acetyl-CoA is transferred to the four-carbon compound of oxaloacetate \n",
    "                                    to form the six-carbon compound of citrate. In a series of reactions two carbons in \n",
    "                                    citrate are oxidized to CO2 and the reaction pathway supplies NADH for use in the \n",
    "                                    oxidative phosphorylation and other metabolic processes. The pathway also supplies \n",
    "                                    important precursor metabolites including 2-oxoglutarate. At the end of the cycle the \n",
    "                                    remaining four-carbon part is transformed back to oxaloacetate. According to the \n",
    "                                    genome sequence data, many organisms seem to lack genes for the full cycle, but \n",
    "                                    contain genes for specific segments.')\n",
    "                                    ;\"\"\")\n",
    "\n",
    "c.execute(\"\"\"INSERT INTO pathways (name,\n",
    "                                    description)\n",
    "                            VALUES\n",
    "                                    ('Pentose Phosphate',\n",
    "                                    'The pentose phosphate pathway is a process of glucose turnover that produces NADPH \n",
    "                                    as reducing equivalents and pentoses as essential parts of nucleotides. There are two \n",
    "                                    different phases in the pathway. One is irreversible oxidative phase in which \n",
    "                                    glucose-6P is converted to ribulose-5P by oxidative decarboxylation, and NADPH is \n",
    "                                    generated. The other is reversible non-oxidative phase in which phosphorylated sugars \n",
    "                                    are interconverted to generate xylulose-5P, ribulose-5P, and ribose-5P. \n",
    "                                    Phosphoribosyl pyrophosphate (PRPP) formed from ribose-5P is an activated compound \n",
    "                                    used in the biosynthesis of histidine and purine/pyrimidine nucleotides. This pathway \n",
    "                                    map also shows the Entner-Doudoroff pathway where 6-P-gluconate is dehydrated and then\n",
    "                                    cleaved into pyruvate and glyceraldehyde-3P.')\n",
    "                                    ;\"\"\")\n",
    "\n",
    "\n",
    "\n",
    "# The other tables are populated automatically by extracting the data from Entrez\n",
    "for (path,enz) in pwt.items():\n",
    "    # Order is the number of the enzyme in the pathway\n",
    "    order = 1\n",
    "    for e in enz:\n",
    "        # The asscoiated table linking pathways and enzymes is created\n",
    "        c.execute(\"\"\"INSERT INTO pathwayenzyme (pathway,\n",
    "                                                enzyme,\n",
    "                                                pathway_order)\n",
    "                                        VALUES\n",
    "                                                (?,\n",
    "                                                ?,\n",
    "                                                ?);\"\"\" ,\n",
    "                                                (path,e,order))\n",
    "        order +=1\n",
    "             \n",
    "       \n",
    "        \n",
    "        # The organism specific sequence is determined here\n",
    "        for o in org:\n",
    "            # The current search is printed to give the user an indication of progress\n",
    "            print('Pathway: ' + path + '; EC: ' + e + '; Organism: ' + o)\n",
    "\n",
    "             # The corresponding protein information is searched for using Entrez\n",
    "            handle = Entrez.esearch(db = 'protein',\n",
    "                                    term = 'EC: ' + e + ' AND ' + o,\n",
    "                                    sort = 'relevance',\n",
    "                                    idtype = 'acc',\n",
    "                                    retmax = 1)\n",
    "\n",
    "            for i in Entrez.read(handle)['IdList']:\n",
    "                handle = Entrez.efetch(db = 'protein', id = i, rettype = 'gb', retmode = 'text')\n",
    "                record = SeqIO.read(handle, \"genbank\")\n",
    "                handle.close()\n",
    "                c.execute(\"\"\"INSERT OR REPLACE INTO enzymes (name,\n",
    "                                                description,\n",
    "                                                ec)\n",
    "                                        VALUES\n",
    "                                                (?,\n",
    "                                                ?,\n",
    "                                                ?);\"\"\" ,\n",
    "                                                (record.name,record.description,e))\n",
    "                \n",
    "            # The corresponding nucleotide sequence information is searched for using Entrez\n",
    "            # The sequence is found using the gene database, the rest of the information is found using the nucleotide one\n",
    "            handle = Entrez.esearch(db = 'gene',\n",
    "                                    term = e + ' ' + o,\n",
    "                                    sort = 'relevance',\n",
    "                                    idtype = 'acc',\n",
    "                                    retmax = 1)\n",
    "            \n",
    "            \n",
    "            for i in Entrez.read(handle)['IdList']:\n",
    "                handle = Entrez.efetch(db = 'nucleotide', id = i, rettype = 'gb', retmode = 'text')\n",
    "                record = SeqIO.read(handle, \"genbank\")\n",
    "                handle.close()\n",
    "\n",
    "                c.execute(\"\"\"INSERT INTO genes (name,\n",
    "                                                organism,\n",
    "                                                sequence)\n",
    "                                        VALUES\n",
    "                                                (?,\n",
    "                                                ?,\n",
    "                                                ?);\"\"\" ,\n",
    "                                                (record.id, o, str(record.seq)))\n",
    "        \n",
    "                c.execute(\"\"\"INSERT INTO enzymegene (ec,\n",
    "                                                    organism,\n",
    "                                                    gene)\n",
    "                                            VALUES\n",
    "                                                    (?,\n",
    "                                                    ?,\n",
    "                                                    ?);\"\"\" ,\n",
    "                                                    (e,o,record.id))\n",
    "            \n",
    "\n",
    "# Update tables\n",
    "conn.commit()\n",
    "\n",
    "# Print tables\n",
    "print(\"\\nPathways table:\")\n",
    "print (pd.read_sql_query(\"SELECT * FROM pathways\", conn))\n",
    "print(\"\\nGenes table:\")\n",
    "print (pd.read_sql_query(\"SELECT * FROM genes\", conn))\n",
    "print(\"\\nEnzymes table:\")\n",
    "print (pd.read_sql_query(\"SELECT * FROM enzymes\", conn))\n",
    "print(\"\\nPathway enzyme associated table:\")\n",
    "print (pd.read_sql_query(\"SELECT * FROM pathwayenzyme\", conn))\n",
    "print(\"\\nEnzyme genes associated table:\")\n",
    "print (pd.read_sql_query(\"SELECT * FROM enzymegene\", conn))\n",
    "                  \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
